### DISCUSSION TITLE
Out of Filament Sensor Monitor for FlashForge Dreamer

### DISCUSSION THREAD

**Discussion by DenPrint** [moonglow/FlashForge_Marlin#38] [Ideas]
Hello,
Great job making Marlin running on FlashForge Dreamer finally i can use Cura.

My FF Dreamer does have filament detection on the Z(-) limit switch with the original FlashForge firmware.
Original on the FF Dreamer there is nothing connected to the Z(-).
Some people put a sensor on there selfs because the option is in the original FF firmware, like this : https://www.ebay.com/itm/281999747282
![IMG_0620_2](https://user-images.githubusercontent.com/86926224/124395543-072e9380-dd05-11eb-8e62-399943937f3e.jpg)

I solderd 4 pins on the open Z(-) connections.
and i have purchased a extra original FF limit switch + FF cable and connected it to open Z(-) on the board.
The new switch do i have connected to a industrial 24Volt timer that gives one shot pulse  of ± 1 second  when running out of filament.
For the original FF firmware this has to be oneshot pulse, otherwise the z axis will not move down when pause the print.

Thanks,

**!!! the Z(-) most be Z(+) see discussion below !!!**

**Comment by moonglow** [comment]
@DenPrint Hi !

Are you sure what they use Z(-) switch ? because Dreamer firmware use Z(-) for homing purpose. I think we can use Z(+) or X(-) Y(-) limit switch for that. As example Dreamer motherboard uses for homing: X max, Y max, Z min all other input is unused and we can use one of them. It be OK if i build test version what will use X(-) switch as example ?

<details><summary>Old Dreamer motherboard:</summary>

![underside](https://user-images.githubusercontent.com/1655387/124396401-f46e8b80-dd11-11eb-91db-1390dae30f90.jpg)

</details>
<details><summary>New Dreamer motherboard ( the same, but also has a dedicated pin for filament detect ):</summary>

![1615465953849](https://user-images.githubusercontent.com/1655387/124396341-a22d6a80-dd11-11eb-8efb-c25199a2709c.png)

</details>

**Reply by DenPrint** [reply]
Hello,

Sorry mistake,  you're right.
i have it connected to the  **Z(+)** connector.       Not the Z(-)   
Z(-) is used for homing.
You can use Z(+) that is also used by FlashForge them self at least for my board version.
![IMG_0622](https://user-images.githubusercontent.com/86926224/124397818-78c10e80-dd12-11eb-940c-52c44416a161.JPEG)


Thanks,

**Reply by DenPrint** [reply]
This is how it works with the original FlashForge Firmware v2.15 20200917 , with the switchs connected to  Z_Max
https://youtu.be/ZknI9RFLSvk

**Reply by moonglow** [reply]
Lets try to check it :) but i do not know what is active polarity of sensor .. so i builded two version with active LOW ( `low` postfix in filename i recommend it first ) and active HIGH. By default **runout sensor is disabled and you need to enable it before using. You can enable it in GUI settings**. Marlin use such script by default if system detect filament sensor was triggered   `#define FILAMENT_RUNOUT_SCRIPT "M600"`. Marlin also supports up to 8 runout sensors ... so it can be actual for dual head machine


[fw_2.0.9.1_filament.zip](https://github.com/moonglow/FlashForge_Marlin/files/6762570/fw_2.0.9.1_filament.zip)


this is notes from Marlin configuration file ( **PRINTJOB_TIMER_AUTOSTART** is enabled in our case ):

```
Filament Runout Sensors
Mechanical or opto endstops are used to check for the presence of filament.

IMPORTANT: Runout will only trigger if Marlin is aware that a print job is running.
Marlin knows a print job is running when:
1. Running a print job from media started with M24.
2. The Print Job Timer has been started with M75.
3. The heaters were turned on and PRINTJOB_TIMER_AUTOSTART is enabled.

```

You can also use `M119` to get current states of endstops, it will output something like that:
```
Reporting endstop status
x_min: open
y_min: open
z_min: TRIGGERED
z_probe: open
filament: open

```

p.s: looks like your chamber temp sensor is dead... or not connected

**Reply by DenPrint** [reply]
I did a first test and it looks good so far i used the:  DREAMER_HIGHT.BIN
See this :  https://youtu.be/G5GT2pO0FRA

Next week i will do some more tests, i first have to setup a good CURA profile.
( Some tips how to set the Start/End G-codes for dual extrusion in CURA for this FF_Marlin?? ).

I also try Marlin on my Old FF Dreamer the  3,5' TFT V1.1  did not work, just show 1/3 of the screen on the left side and mirrored.

Thanks,

**Reply by moonglow** [reply]
Wow... looks nice !) Thanx for video !

>>I also try Marlin on my Old FF Dreamer the 3,5' TFT V1.1 did not work, just show 1/3 of the screen on the left side and mirrored.

Hm.. interesting situation, can you provide me some photos how it looks like, i think this is good old Dreamer for my colection of Dreamer variants :) Your printer usable with Marlin but only via USB port, of course it is not good and i think i need to add such variant to my Marlin port if you can assist me to give new life to old good Dreamer.

>Some tips how to set the Start/End G-codes for dual extrusion in CURA for this FF_Marlin??

FF Marlin is regular Marlin firmware, so we can just select some standard Cura printer profile which was similar to ours and use they G-start code as our base for other modification. As example dualhead `BIBO2 dual`

Start G-Code ( only for reference ! ):
```
M104 T0 165
M104 T1 165
M109 T{initial_extruder_nr} S{material_print_temperature_layer_0, initial_extruder_nr}
G21 ;metric values
G90 ;absolute positioning
M107 ;start with the fan off
G28 X0 Y0 ;move X/Y to min endstops
G28 Z0 ;move Z to min endstops
G1 Z2.0 F400 ;move the platform down 2mm
T0
G92 E0
G28
G1 Y0 F1200 E0
G92 E0
T{initial_extruder_nr}
M117 BIBO Printing...
```
End G-Code
```
;End GCode
M104 T0 S0                     ;extruder heater off
M104 T1 S0                     ;extruder heater off
M140 S0                     ;heated bed heater off (if you have it)
G91
G1 Z1 F100                                        ;relative positioning
G1 E-1 F300                            ;retract the filament a bit before lifting the nozzle, to release some of the pressure
G1 Z+0.5 E-2 X-20 Y-20 F300 ;move Z up a bit and retract filament even more
G28 X0 Y0                              ;move X/Y to min endstops, so the head is out of the way
M84                         ;steppers off
G90                         ;absolute positioning
```

**Reply by DenPrint** [reply]
Thanks,

I have made a new dissusion about the TFT V1.1

_# p.s: looks like your chamber temp sensor is dead... or not connected_
I bought this Dreamer 2nd hand, i think the previous owner removed it.
Or is this old version of the  Dreamer also sold without this sensor installed??

**Reply by moonglow** [reply]
>Or is this old version of the Dreamer also sold without this sensor installed??

I think it can be installed by yourself because it is regular cheap thermistor for 100k ( looks like water drop ) connected to 2 pins ( white socket XH 2.54 ) , as example you can use something like that: [NTC_100K](https://aliexpress.com/item/32877679034.html)

**Reply by DenPrint** [reply]
I have try to make my own firmware in Visual Studio Code with Marlin Auto Build, to att a second filament sensor.
But when load the new firmware over usb it does not want to boot.
Do i have to use the FlashForge firmware tool? where to find the .exe and how to use it?

Thanks,

**Reply by DenPrint** [reply]
I found the tool https://github.com/moonglow/FlashForge_Marlin/issues/18 but how do i use it with Windows10?

**Reply by moonglow** [reply]
>Do i have to use the FlashForge firmware tool? where to find the .exe and how to use it?

Yep ! You are right ! You need to encrypt your firmware before flash it to printer ( printer bootloader accept only encrypted firmwares and try to decrypt it first, if you provide raw binary it will decrypted to shuffled binary data )

>I found the tool #18 but how do i use it with Windows10?

Official repo here: [LINK](https://github.com/moonglow/flashforge_firmware_tool), to build encryption tool you can use any C compiler ( as example freeware Pelles C compiler ), but if you need prebuild version here is it: [ff_fw_tool.zip](https://github.com/moonglow/FlashForge_Marlin/files/6835208/ff_fw_tool.zip)

>i use it with Windows10?

You need to encrypt `firmware.bin` file, for FF Dreamer you can omit encryption key option ( -k ) and use it like that, ( i assume your firmware file in same folder where is ff tool was located ):

`ff_fw_tool -e -i firmware.bin -o dreamer.bin`

**Reply by DenPrint** [reply]
Thanks i going to try.

**Reply by DenPrint** [reply]
i did try to download the ff_fw_tool.zip but Windows Defender does not like the file inside the zip. and blocks it right away.
instead i compiled the file with Pelles C.
![test01](https://user-images.githubusercontent.com/86926224/126062395-628be546-05be-4fe9-a87d-01734dd09e1c.PNG)
But is this the right way to start it?

**Reply by moonglow** [reply]
>but Windows Defender does not like the file inside the zip

o_O you can try to upload to virustotal  ( VT report [LINK](https://www.virustotal.com/gui/file/1d02f3421f99be0e774814db6cd46f40f049c13137cbfa88d3ff57c5ed506c40/detection) ) looks like only paranoiac Microsoft interpreted it like Trojan, but i never publish here ( and in other places too ) viruses or other software for destruction ^_^

>But is this the right way to start it?

You can use example from my previous message, but i think i need to change description.... [] - mean what you can omit this parameter, `<key>` mean key value, no need to use `<>`.. Some examples ( both of them correct ):


`ff_fw_tool -e -i firmware.bin -o dreamer.bin` - key option omitted
`ff_fw_tool -k flashforge790315 -e -i firmware.bin -o dreamer.bin` - with  key option provided

**Reply by DenPrint** [reply]
![test02](https://user-images.githubusercontent.com/86926224/126064765-a07de608-2ad8-4c90-9b8a-bcdd6d54ff2c.PNG)

That did the trick.
Now i can do some more testing.

Thanks,
